In [13]:
import pandas as pd
from sklearn import linear_model
from sklearn.metrics import r2_score

In [14]:
df = pd.read_csv('housing.csv')

Elimina los máximos

In [18]:
df.median_house_value.max()

500000.0

In [19]:
df= df[df.median_house_value != df.median_house_value.max()]

In [20]:
df.medi

AttributeError: 'DataFrame' object has no attribute 'medi'

## Crea un modelo de regresión lineal simple usando el ingreso y calcula su R2

In [29]:
model_income= linear_model.LinearRegression()
X_income= df[['median_income']]
target= 'median_house_value'
y= df[target]
model_income.fit(X_income,y)
predictions_income= model_income.predict(X_income)   
r2_score(y_pred=predictions_income, y_true=y)

0.41824601638696535

In [30]:
import numpy as np
np.var(y)

9430038184.981813

In [31]:
np.var(y - predictions_income)

5485962279.736201

## Crea un modelo de regresión lineal simple usando la edad de las casas

In [32]:
model_age = linear_model.LinearRegression()
X_age = df[['housing_median_age']]

model_age.fit(X_age, y)
predictions_age = model_age.predict(X_age)
r2_score(y_pred=predictions_age, y_true=y)

0.004243086267057494

## Crea un modelo de regresión lineal multiple usando las variables anteriores

In [35]:
X_multiple= df[['housing_median_age', 'median_income']]
model_multiple= linear_model.LinearRegression()
model_multiple.fit(X_multiple, y)
predictions_multiple= model_multiple.predict(X_multiple)
r2_score(y_true=y, y_pred=predictions_multiple)


0.45641609987943765

Escribe la ecuación del modelo

In [36]:
# Interepto
model_multiple.intercept_

-7706.582194528077

In [37]:
# Coeficientes
model_multiple.coef_

array([ 1547.19322302, 42395.50726961])

$$Precio = \beta_0 + \beta_1 * x_1 + \beta_2* x_2$$

$$Precio = -7706.58 + 1547.19 * edad + 42395.5072* ingreso$$

### Agrega la variable de households

In [44]:
X_multiple2= df[['housing_median_age', 'median_income','households']]
model_multiple2= linear_model.LinearRegression()
model_multiple2.fit(X_multiple2, y)
predictions_multiple2= model_multiple2.predict(X_multiple2)
r2_score(y_true=y, y_pred=predictions_multiple2)

0.4738652646118531

Escribe la ecuación del modelo

In [45]:
model_multiple2.intercept_

-35381.918921855715

In [46]:
model_multiple2.coef_

array([1.88443277e+03, 4.25229830e+04, 3.51923940e+01])

### Agrega todas

In [ ]:
### Verific los NA's

import pandas as pd

X = df.copy()
X = X.drop(target, axis=1)

X = pd.get_dummies(X)



In [62]:
# Quita NA's 
df.isna().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        200
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [65]:
X['total_bedrooms']= X.total_bedrooms.fillna(df.total_bedrooms.mean())

In [66]:
model_todo = linear_model.LinearRegression().fit(X, y)

In [67]:
# Convertir a one hot encoding 
predictions_todo = model_todo.predict(X)

In [70]:
r2_score(y_pred= predictions_todo, y_true= y)

0.6160654814263344

In [71]:
model_todo.intercept_

-2013400.6369967817

In [75]:
coeficientes= pd.DataFrame({
    'names': X.keys(),
    'coeficientes': model_todo.coef_
})
coeficientes

names   coeficientes
0                    longitude  -23888.866252
1                     latitude  -22035.549806
2           housing_median_age     898.336237
3                  total_rooms      -5.629652
4               total_bedrooms      64.348693
5                   population     -31.141217
6                   households      67.327026
7                median_income   37842.829940
8    ocean_proximity_<1H OCEAN  -25370.421296
9       ocean_proximity_INLAND  -64603.345648
10      ocean_proximity_ISLAND  143605.492001
11    ocean_proximity_NEAR BAY  -32119.044648
12  ocean_proximity_NEAR OCEAN  -21512.680409

### Visualizando el modelo, calibracion

### Compara la calibración de todos los modelos anteriores